src/audio/pipeline.c

```diff
@@ -145,26 +145,56 @@ struct pipeline_walk_context {
 	/**< pipelines to be scheduled after trigger walk */
 	struct list_item pipelines;
 	/*
-	 * If this flag is set, pipeline_for_each_comp() will skip all
+	 * If this flag is set, pipeline_for_each_element() will skip all
 	 * incompletely initialised components, i.e. those, whose .pipeline ==
 	 * NULL. Such coomponents should not be skipped during initialisation
 	 * and clean up, but they should be skipped during streaming.
 	 */
 	bool skip_incomplete;
+
+	/**
+	 * when true, then comp_func is called after processing component
+	 * buffers and following components, otherwise comp_func is called at
+	 * first place.
+	 *
+	 * Example:
+	 * For topology: c1 -> b2 -> c3 -> b4 -> c5
+	 * context functions will be applied in order:
+	 *   false: c1, b2, c3, b4, c5
+	 *    true: c5, b4, c3, b2, c1
+	 */
```

Comment:
This just looks like we are changing the direction of travel ? We should really use terms like "downstream" to describe source -> sink and "upstream" to describe "sink -> source" calling direction (i.e downstream is the direction of travel for audio samples).

Reply:
Please consider following topology:
```
c1 -> b1 -+
          |-> c3 -> b3 -> c4
c2 -> b2 -+
```

When we start from `c1` and go downstream with `call_func_on_exit` set to true (c4->b3->c3->b1->c1) then it's not the same as starting  from `c4` and going upstream with  `call_func_on_exit` set to false (then `c2` an `b2` also will be proceeded). This mechanism is already implicit used during (free, hw params, params neg, copy) propagation, it's just explicit form of it. I would like to keep pipe walk behavior as it is here, (only stack usage is reduced), to reduce possibility of bug introduction.

For example, `pipeline_free()` goes downstream, and for each component removes connection with sink buffers, so calling this function as it is on component exit in `pipeline_for_each_element()` is safer than calling it on component entry and then iterating to the next component via disconnected buffers.

Reply:
> Please consider following topology:
> 
> ```
> c1 -> b1 -+
>           |-> c3 -> b3 -> c4
> c2 -> b2 -+
> ```
> 
> When we start from `c1` and go downstream with `call_func_on_exit` set to true (c4->b3->c3->b1->c1) then it's not the same as starting from `c4` and going upstream with `call_func_on_exit` set to false (then `c2` an `b2` also will be proceeded).

Sorry, you've lost me. Wouldn't c1 downstream be ```c1 -> b1 -> c3 ->b3 -> c4``` and c2 downstream be ```c2 -> b2 -> c3 -> b3-> c4``` ?

c4 upstream would be ```c4 -> b3 -> c3 -> b1 -> c1 ->b2 -> c2```

So what does ```call_func_on_exit``` mean or do.

When do we need to travel upstream ?

When do we need to travel downstream ?

Reply:
> Sorry, you've lost me. Wouldn't c1 downstream be `c1 -> b1 -> c3 ->b3 -> c4` and c2 downstream be` c2 -> b2 -> c3 -> b3-> c4` ?

That's right. Travel direction + starting node defines components to enter, `call_func_on_exit` can just reverse entry order (entered component list is kept unchanged), so maybe I rename it to `reverse_order`.

Reply:
ok, so the naming of  ```call_func_on_exit``` describes the internals of the implementation but not the usage.

IIUC our use case is we want to walk c4 -> c1, but c1 is our entry component (i.e. c4 is unknown to us at this point) ?  It would be better here to have 

1) for_each_downstream_endpoint(entry, conditions, callback)

2) for_each_upstream_endpoint(entry, conditions, callback)

As the logic is becoming more complex in the recursive function deciding whether to "call func on exit".

So in the example, we call 1) and our callback constructs a list of endpoints (in this case it's just c4). This list of endpoints is then used in for_each_component(c4, upstream, some_func).

This keeps everything simple and avoids complexity in the recursive logic.